sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,162 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/freestanding/limits.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_aon_timer.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_rv_timer.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/aon_timer_testutils.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+static const uint32_t kTickFreqHz = 1000 * 1000;  // 1Mhz / 1us
+static const uint32_t kSleepTime_us = 10 * 1000;  // 10ms
+static dif_aon_timer_t aon_timer;
+static dif_rv_timer_t rv_timer;
+static dif_rv_plic_t plic;
+
+static volatile dif_aon_timer_irq_t irq;
+static volatile top_earlgrey_plic_peripheral_t peripheral;
+static volatile uint64_t time_elapsed;
+
+/** Initialize the rv timer to count the tick
+ */
+static void tick_init(void) {
+  CHECK_DIF_OK(dif_rv_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR), &rv_timer));
+
+  CHECK_DIF_OK(dif_rv_timer_reset(&rv_timer));
+
+  // Compute and set tick parameters (i.e., step, prescale, etc.).
+  dif_rv_timer_tick_params_t tick_params;
+  CHECK_DIF_OK(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz,
+                                                    kTickFreqHz, &tick_params));
+
+  CHECK_DIF_OK(
+      dif_rv_timer_set_tick_params(&rv_timer, kPlicTarget, tick_params));
+
+  CHECK_DIF_OK(dif_rv_timer_counter_set_enabled(&rv_timer, kPlicTarget,
+                                                kDifToggleEnabled));
+}
+
+/** Read the current rv timer count/tick
+ * @return The current rv timer count.
+ */
+static uint64_t tick_count_get(void) {
+  uint64_t tick = 0;
+  dif_rv_timer_counter_read(&rv_timer, kPlicTarget, &tick);
+  return tick;
+}
+
+/** Execute the wake up interrupt test
+ */
+static void execute_test(dif_aon_timer_t *aon_timer, uint32_t sleep_time_us) {
+  peripheral = kTopEarlgreyPlicPeripheralUnknown;
+  irq = kDifAonTimerIrqWdogTimerBark;
+  // The wake up time should be `sleepTime_us ±15%`
+  uint32_t sleep_range_h = (float)sleep_time_us * 1.15;
+  uint32_t sleep_range_l = (float)sleep_time_us * 0.85;
```

Comment:
We have `-march=rv32imc`, which means it would use software floating point, not sure we want that. Could you achieve similar effect without float?

Reply:
Yes, is easy to not use `float` here, but would be a problem to have `float`?

Reply:
Well, as it doesn't have hardware instructions, it would have to pull in some sort of software implementation. I am no expert, so better ask @mundaym . But the point being, why do we need to resort something that is not native for our platform if there is no real need for it?

Reply:
You could compile and see with `objdump` what `float` pulls in/does.